src/send/SelectRecipientButtons.tsx

```diff
@@ -0,0 +1,177 @@
+
+const CONTACTS_PERMISSION =
+  Platform.OS === 'ios' ? PERMISSIONS.IOS.CONTACTS : PERMISSIONS.ANDROID.READ_CONTACTS
+
+type Props = {
+  onContactsPermissionGranted: () => void
+}
+
+export default function SelectRecipientButtons({ onContactsPermissionGranted }: Props) {
+  const { t } = useTranslation()
+  const phoneNumberVerified = useSelector(phoneNumberVerifiedSelector)
+  const [contactsPermissionStatus, setContactsPermissionStatus] = useState<
+    PermissionStatus | undefined
+  >(undefined)
+  const [showConnectPhoneNumberModal, setShowConnectPhoneNumberModal] = useState(false)
+  const [showEnableContactsModal, setShowEnableContactsModal] = useState(false)
+  const [navigateToPhoneVerification, setNavigateToPhoneVerification] = useState(false)
+
+  useAsync(async () => {
+    if (!contactsPermissionStatus) {
+      setContactsPermissionStatus(await check(CONTACTS_PERMISSION))
+    }
+  }, [contactsPermissionStatus])
+
+  const onPressContacts = async () => {
+    const currentPermission = await check(CONTACTS_PERMISSION)
+    setContactsPermissionStatus(currentPermission)
+
+    ValoraAnalytics.track(SendEvents.send_select_recipient_contacts, {
+      contactsPermissionStatus: currentPermission,
+      phoneNumberVerified,
+    })
+
+    if (!phoneNumberVerified) {
+      setShowConnectPhoneNumberModal(true)
+      return
+    }
+
+    // Based on https://github.com/zoontek/react-native-permissions#understanding-permission-flow
+    if (currentPermission === RESULTS.BLOCKED) {
+      // permission not requestable
+      setShowEnableContactsModal(true)
+    } else if (currentPermission === RESULTS.GRANTED) {
+      // permission already granted
+      onContactsPermissionGranted()
+    } else if (currentPermission === RESULTS.DENIED) {
+      ValoraAnalytics.track(SendEvents.send_select_recipient_request_contacts_permission_started)
+      const newPermission = await request(CONTACTS_PERMISSION, {
+        // rationale for Android, shows up the 2nd time a permission is requested.
+        title: t('accessContacts.disclosure.title'),
+        message: t('accessContacts.disclosure.body'),
+        buttonPositive: t('continue'),
+        buttonNegative: t('notNow') ?? undefined,
+      })
+      ValoraAnalytics.track(
+        SendEvents.send_select_recipient_request_contacts_permission_completed,
+        { permissionStatus: newPermission }
+      )
+      if (newPermission === RESULTS.GRANTED) {
+        // permission granted
+        onContactsPermissionGranted()
+      } else if (newPermission === RESULTS.BLOCKED && Platform.OS === 'android') {
+        // we only know if permissions are requestable or not on Android after
+        // doing a request. So if we get back blocked from the request, its
+        // possible a native modal was never shown, so show our custom modal.
+        setShowEnableContactsModal(true)
+      }
+    } else {
+      // this should never happen
```

Comment:
it would be nice to get static validation that it won't happen. one clever way I've seen of making this happen is using a switch to cover all the cases, and then in the default assign newPermission to a variable with type never. Then if we update a dependency and there's a new member of the RESULTS enum, we get a compilation error

Reply:
https://github.com/valora-inc/wallet/blob/5014be624ff3a626dba17f433701e2bde21619f4/src/swap/SwapScreen.tsx#L296

Reply:
There are actually 2 other potential values, `limited` and `unavailable`, which I believe should never happen for contacts permission and this else was to mainly cover that and not new states.. I guess I can change to a switch case to get compilation error on new states.

Reply:
updated to use switch case with type check